File: plugins/bigtiny/bigtiny/server/routes/recipes.py
from __future__ import annotations

import json
import logging
from uuid import uuid4

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

from bigtiny.recipes.engine import RecipeEngine

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/recipes")


class CreateRecipeRequest(BaseModel):
    name: str
    prompt_template: str
    instructions: str | None = None
    parameters: list[str] | None = None
    required_mcp_servers: list[str] | None = None
    system_prompt_layer: str | None = None
    max_steps: int = 30


class ExecuteRecipeRequest(BaseModel):
    parameters: dict = {}


@router.get("")
async def list_recipes(request: Request):
    db = request.app.state.db
    rows = await db.fetch_all("SELECT * FROM recipes")
    return {"recipes": [dict(r) for r in rows]}


@router.post("")
async def create_recipe(body: CreateRecipeRequest, request: Request):
    db = request.app.state.db
    recipe_id = uuid4().hex[:8]
    await db.execute(
        "INSERT INTO recipes (id, name, prompt_template, instructions, parameters, "
        "required_mcp_servers, system_prompt_layer, max_steps) "
        "VALUES (:id, :name, :prompt, :instr, :params, :servers, :layer, :steps)",
        {
            "id": recipe_id,
            "name": body.name,
            "prompt": body.prompt_template,
            "instr": body.instructions,
            "params": json.dumps(body.parameters or []),
            "servers": json.dumps(body.required_mcp_servers or []),
            "layer": body.system_prompt_layer,
            "steps": body.max_steps,
        },
    )
    return {"id": recipe_id, "status": "created"}


@router.post("/{recipe_id}/execute")
async def execute_recipe(recipe_id: str, body: ExecuteRecipeRequest, request: Request):
    recipe_engine: RecipeEngine = request.app.state.recipe_engine
    try:
        session_id = await recipe_engine.execute(recipe_id, body.parameters)
        return {"session_id": session_id}
    except ValueError as e:
        raise HTTPException(404, detail=str(e))
    except Exception as e:
        raise HTTPException(500, detail=str(e))


@router.delete("/{recipe_id}")
async def delete_recipe(recipe_id: str, request: Request):
    db = request.app.state.db
    existing = await db.fetch_one(
        "SELECT * FROM recipes WHERE id = :id", {"id": recipe_id}
    )
    if not existing:
        raise HTTPException(404, "Recipe not found")
    await db.execute("DELETE FROM recipes WHERE id = :id", {"id": recipe_id})
    return {"status": "deleted"}

File: BigTinyV2/daemon/src/routes/recipes.rs
//! `/api/recipes` routes — mirrors
//! `plugins/bigtiny/bigtiny/server/routes/recipes.py`.

use std::sync::Arc;

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::{Extension, Json};
use serde_json::{json, Value};

use crate::error::RecipeError;
use crate::storage::apps::AppIdentity;
use crate::storage::recipes;

use super::AppState;

fn err_response(status: StatusCode, message: impl Into<String>) -> Response {
    (status, Json(json!({"error": message.into()}))).into_response()
}

pub async fn list_recipes(
    State(state): State<Arc<AppState>>,
    Extension(identity): Extension<AppIdentity>,
) -> Response {
    match recipes::list_recipes_for_app(&state.db, &identity.app_id).await {
        Ok(rows) => Json(json!({"recipes": rows})).into_response(),
        Err(e) => err_response(StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
    }
}

pub async fn create_recipe(
    State(state): State<Arc<AppState>>,
    Extension(identity): Extension<AppIdentity>,
    Json(body): Json<Value>,
) -> Response {
    let (Some(name), Some(prompt_template)) = (
        body.get("name").and_then(|v| v.as_str()),
        body.get("prompt_template").and_then(|v| v.as_str()),
    ) else {
        return err_response(
            StatusCode::BAD_REQUEST,
            "name and prompt_template are required",
        );
    };
    let instructions = body.get("instructions").and_then(|v| v.as_str());
    let max_steps = body.get("max_steps").and_then(|v| v.as_i64()).unwrap_or(30) as i32;

    let id = uuid::Uuid::new_v4().to_string();
    match recipes::create_recipe(
        &state.db,
        &id,
        name,
        prompt_template,
        instructions,
        max_steps,
        &identity.app_id,
    )
    .await
    {
        Ok(()) => Json(json!({"id": id})).into_response(),
        Err(e) => err_response(StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
    }
}

pub async fn delete_recipe(
    State(state): State<Arc<AppState>>,
    Extension(identity): Extension<AppIdentity>,
    Path(id): Path<String>,
) -> Response {
    match recipes::delete_recipe_for_app(&state.db, &id, &identity.app_id).await {
        // Not ours, or gone: a 404. Reporting success for a delete that
        // matched nothing makes a client drop the row from its list and then
        // watch it reappear on the next refresh.
        Ok(0) => err_response(StatusCode::NOT_FOUND, format!("no such recipe: {id}")),
        Ok(_) => Json(json!({"ok": true})).into_response(),
        Err(e) => err_response(StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
    }
}

pub async fn execute_recipe(
    State(state): State<Arc<AppState>>,
    Extension(identity): Extension<AppIdentity>,
    Path(id): Path<String>,
    body: Option<Json<Value>>,
) -> Response {
    // Executing another app's recipe would run its workflow on its
    // provider and against its billing account.
    match recipes::get_recipe_for_app(&state.db, &id, &identity.app_id).await {
        Ok(Some(_)) => {}
        Ok(None) => return err_response(StatusCode::NOT_FOUND, "recipe not found"),
        Err(e) => return err_response(StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
    }
    let parameters = body.map(|Json(b)| b).unwrap_or_else(|| json!({}));
    match state.recipe_engine.execute(&id, parameters).await {
        Ok(session_id) => Json(json!({"session_id": session_id})).into_response(),
        Err(e) => {
            // Distinguish validation/usage errors (missing recipe → 404,
            // bad template → 400) from internal failures (storage → 500,
            // provider-failed turn → 500). The previous code mapped *every*
            // engine error to 400, hiding genuine server faults as client
            // errors.
            let status = match &e {
                RecipeError::NotFound(_) => StatusCode::NOT_FOUND,
                RecipeError::Template(_) => StatusCode::BAD_REQUEST,
                RecipeError::Storage(_) | RecipeError::TurnFailed(_) => {
                    StatusCode::INTERNAL_SERVER_ERROR
                }
            };
            err_response(status, e.to_string())
        }
    }
}
